pennylane/pauli/pauli_arithmetic.py

```diff
@@ -414,10 +440,61 @@ def _pw_wires(w: Iterable) -> wires.Wires:
 
         return math.expand_matrix(reduced_mat, result_wire_order, wire_order=wire_order)
 
+    def _to_sparse_mat(self, wire_order):
+        """Compute the sparse matrix of the Pauli sentence by efficiently adding the Pauli words
+        that conform it. See pauli_sparse_matrices.md for the technical details."""
+        pauli_words = list(self)  # Ensure consistent ordering
+        n_wires = len(wire_order)
+        matrix_size = 2**n_wires
+        matrix = sparse.csr_matrix((matrix_size, matrix_size), dtype="complex128")
+        op_sparse_idx = _ps_to_sparse_index(pauli_words, wire_order)
+        _, unique_inds, unique_invs = np.unique(
+            op_sparse_idx, axis=0, return_index=True, return_inverse=True
+        )
+        pw_sparse_structures = unique_inds[unique_invs]
+
+        buffer_size = 2 ** max(0, 31 - n_wires - 5)  # Buffer of 1GB memory
+        mat_data = np.empty((matrix_size, buffer_size), dtype=np.complex128)
+        mat_indices = np.empty((matrix_size, buffer_size), dtype=np.int64)
+        i = 0
+        for sparse_structure in unique_inds:
+            idx = np.where(pw_sparse_structures == sparse_structure)[0]
+            mat = self._sum_same_structure_pws([pauli_words[i] for i in idx], wire_order)
+            mat_data[:, i] = mat.data
+            mat_indices[:, i] = mat.indices
+
+            i += 1
+            if i == buffer_size:
+                # Add partial results in batches to control the memory usage
+                matrix += self._sum_different_structure_pws(mat_indices, mat_data)
+                i = 0
+
+        return matrix + self._sum_different_structure_pws(mat_indices[:, :i], mat_data[:, :i])
+
+    def _sum_same_structure_pws(self, pauli_words, wire_order):
+        """Sums Pauli words with the same sparse structure."""
+        mat = pauli_words[0].to_mat(wire_order, coeff=self[pauli_words[0]], format="csr")
+        for word in pauli_words[1:]:
+            mat.data += word.to_mat(wire_order, coeff=self[word], format="csr").data
+        return mat
+
+    @staticmethod
+    def _sum_different_structure_pws(indices, data):
+        """Sums Pauli words with different parse structures."""
+        size = indices.shape[0]
+        idx = np.argsort(indices, axis=1)
+        matrix = sparse.csr_matrix((size, size), dtype="complex128")
+        matrix.indices = np.take_along_axis(indices, idx, axis=1).ravel()
+        matrix.data = np.take_along_axis(data, idx, axis=1).ravel()
+        matrix.indptr = _cached_arange(size + 1) * indices.shape[1]
+        matrix.data[np.abs(matrix.data) < 1e-8] = 0  # Faster than np.isclose(matrix.data, 0)
```

Comment:
The dynamic range for `np.complex128` is `np.finfo(np.complex128).eps` which is usually around `1e-16`. Assuming the entries are of order 1, I would use that for truncation.

Reply:
I tried setting it to 1e-16 but then there are some elements of the order of 1e-10 that are not removed. However, if we used `np.isclose(x, 0)` those would be eliminated. Usually I find about 500 unique sparse structures among ~3000 Pauli words which make every entry to sum about 60 terms. So expecting the entries to be of order 1-10 would be reasonable though.

Reply:
Oh, I entirely forgot about the coefficients that multiply every Pauli word, this means we can have any arbitrary scale potentially :/ should we maybe stick to `np.isclose`?

Reply:
`np.isclose` is convenient to `assert` in tests, but it is not a good source of truth for what is actually close to 0. Often, if you want an answer precise to ~1e-8, you do need double the precision throughout the calculations (i.e. 1e-16). 

Here, I think we want to truncate entries below `np.finfo(np.complex128).eps` not to lose precision. If not too expensive, you could multiply that by `np.max(np.abs(data))`. Otherwise, I think units are such that this will come out about `1` plus/minus an order of magnitude or two. You could double-check that with the algo squad.